[[Bugfix 19619]] Add color to the clock widget

andrewferguson opened the pull request:
I've added several new properties to the clock widget to allow for the colors of the various components (numbers, second hand, minute hand, hour hand, face) to be changed. In total there are ten new properties: five for the day and five for night.

I did consider trying to use existing LiveCode properties such as the backgroundColor for the face color and the foregroundColor for the number color, but this was complicated by the clock having day and night modes which meant that either the colors would be inverted at night (potentially leading to an ugly appearance - particularly because the night mode cannot be disabled) or a separate set of properties would be created for the night. In the end I decided on going for a whole new set of properties that could be named consistently.

I've updated both the tests and the documentation for the clock widget to include these new properties.

andrewferguson committed:
[[Bugfix 19619]] Added properties <dayNumberColor> and <nightNumberColor> to clock widget

The properies <dayNumberColor> and <nightNumberColor> have been added to the
clock widget to allow the colors of the numbers on the face of the clock to
be changed and for a different color to be displayed at night time (as set
my the <isDay> read-only property).

andrewferguson committed:
[[Bugfix 19619]] Added <dayHourHandColor> and <nightHourHandColor> to clock widget

The properies <dayHourHandColor> and <nightHourHandColor> have been added to the
clock widget to allow the colors of the hour hand of the clock to
be changed and for a different color to be displayed at night time (as set
my the <isDay> read-only property).

andrewferguson committed:
[[Bug 19619]] Fix minor typos

Fixed two typographical errors:

   1.) The documentation for the <nightHourHandColor> referred to the
       previous <nightNumberColor> property

   2.) The default <nightNumberColor> property for older widgets that
       did not have the property is set to black (the day color) rather
       than white (the default night color)

andrewferguson committed:
[[Bugfix 19619]] Added <dayMinuteHandColor> and <nightMinuteHandColor> to clock widget

The properies <dayMinuteHandColor> and <nightMinuteHandColor> have been added to the
clock widget to allow the colors of the minute hand of the clock to
be changed and for a different color to be displayed at night time (as set
by the <isDay> read-only property).

andrewferguson committed:
[[Bugfix 19619]] Added <daySecondHandColor> and <nightSecondHandColor> to clock widget

The properies <daySecondHandColor> and <nightSecondHandColor> have been added to the
clock widget to allow the colors of the second hand of the clock to
be changed and for a different color to be displayed at night time (as set
by the <isDay> read-only property).

andrewferguson committed:
[[Bugfix 19619]] Added properties <dayFaceColor> and <nightColorColor> to clock widget

The properies <dayFaceColor> and <nightFaceColor> have been added to the
clock widget to allow the color of the face of the clock to
be changed and for a different color to be displayed at night time (as set
by the <isDay> read-only property).

andrewferguson committed:
[[Bugfix 19619]] The outer nub inherits its color from the minute hand

The minute hand of the clock is the first hand to be drawn, directly above
the outer nub. In the same way that the inner nub (which is drawn last)
inherits the color of the second hand (which is drawn second-last),  the
outer nub should probably inherit the color of the minute hand.

andrewferguson committed:
[[Bugfix 19619]] Updated overview documentation and version number

The widget overview documnetation has been updated to include information
abou the new properties that allow for custom styling of the clock.

The version number of the widget has been increased by 0.1, in-line with
Semantic Versioning (the property changes are backwards compatible, older
clock widgets can be imported and will continue to function as they did
before).

andrewferguson committed:
[[Bugfix 19619]] Added release note and updated user guide

The user guide for the clock has been updated to reflect the new set
of properties that the clock widget has for changing its colors.

A release note has been added.

andrewferguson committed:
[[Bugfix 19619]] Added tests for new properies of clock widget

The test file 'properties.livecodescript' has been updated to include
tests for all of the newly-added properties for changing the day and
night colors of the clock.

All these tests have been run and pass OK, so hopefully no issues here.

livecodeali commented on extensions/widgets/clock/clock.lcb:
```diff
@@ -99,10 +295,72 @@ public handler OnSave(out rProperties as Array)
+		put stringToColor(pProperties["dayNumberColor"]) into mDayNumberColor
+	else
+		put color kBlack into mDayNumberColor
+	end if
```
You don't need these `else` clauses - `OnCreate` is always called before `OnLoad` so the defaults will already be set there

andrewferguson replied:
Ah, neat. I thought it only called OnCreate when the widget was first created - it makes more sense that it's when it's created when the stack opens and then all the properties are loaded in.

Will fix.

livecodeali commented on extensions/widgets/clock/clock.lcb:
```diff
@@ -85,6 +271,16 @@ private variable mCurrentTime 	as List
+private variable mDaySecondHandColor as Color
+private variable mNightSecondHandColor as Color
+private variable mDayFaceColor as Color
+private variable mNightFaceColor as Color
```
It might be better to store all the color variables as `String`s instead of `Color`s, and that way you only need to convert to a color in one place (namely when painting) - it simplifies the getters, setters, and `OnLoad`, `OnSave` and `OnCreate` that way.

livecodeali commented on extensions/widgets/clock/clock.lcb:
```diff
@@ -76,6 +80,188 @@ UTC that the clock should display.
+metadata dayNumberColor.editor is "com.livecode.pi.color"
+metadata dayNumberColor.default is "0,0,0"
+metadata dayNumberColor.section is "Colors"
+metadata dayNumberColor.label is "Number color (day)"
```
There's (probably) no reason not to allow the colors to have alpha values - all you'd need to do is use the 
`"com.livecode.pi.colorwithalpha"` editor and make sure all the default color strings have full opacity as the fourth item

andrewferguson replied:
I borrowed that from the colorSwatch widget, so I didn't realise that there was a dedicated PI editor for alpha colors. Agree it's probably best - by default the dayFaceColor is partially transparent.

livecodeali reviewed (commented):
There's nothing actually wrong here - what you've done works perfectly! I just have a few suggestions.

andrewferguson committed:
[[Bugfix 19619]] Removed unnecessary else-if clauses

Removed the else-if clauses that set the property values to their
defaults if the widget is an old (pre 1.1.3) version that does not
have these properties. These property values will have already
been set by the OnCreate handler, so are unneeded.

andrewferguson committed:
[[Bugfix 19619]] Allow for alpha colors in the property inspector

Updated the property metadata definitions for the various color properties
so that they are rendered with a slider for alpha value in the properties
inspector.

andrewferguson committed:
[[Bugfix 19619]] Store color properties in Strings, not Colors

The ten various color properties of the clock widget are now stored in
strings rather than color variables. This simplifies the code by
elimintating the need for seperate getters for all the color properties.
Instead, the conversion from a string to a color is done whenever the
color is required for use in the OnPaint handler.

andrewferguson committed:
[[Bugfix 19619]] Updated test file

The test script for the properties of the clock has been updated
to reflect the new method of storing the color properties as strings.
With this new method, if the color is set to a string that does not
include an alpha value, then no alpha value will be returned.

andrewferguson committed:
[[Bugfix 19619]] Added dependancies for clock to tests

Two of the test files required the clock module in the tests.
The new clock module (version 1.1.3) now requires the widgetutils
library, and if this was not loaded prior to loading the clock
module then the test would fail.

For some reason all these tests run fine on my machine even when
the widgetutils dependancy is not loaded prior to the clock module,
so I cannot confirm 100% that all the tests will run fine this time!

livecodeali commented on extensions/widgets/clock/clock.lcb:
```diff
@@ -33,12 +36,13 @@ use com.livecode.canvas
 metadata title is "Clock"
 metadata author is "LiveCode"
-metadata version is "1.0.3"
+metadata version is "1.1.3"
```
This can probably be just 1.1.0 actually (and in the comment below where you say what version the props were added in)

livecodeali commented on extensions/widgets/clock/notes/19619.md:
```diff
@@ -0,0 +1,4 @@
+---
+version: 9.0.0-dp1
+---
```
The version here will be 9.1.0-dp-1

andrewferguson committed:
[[Bugfix-19619]] Correct version numbers

Two version numbers that were incorrectly set have been updated.The
first is the widget version number, which was erroneously set to
1.1.3 from 1.0.3. It has now been set to 1.1.0, indicating a new
update with no bugfixes (yet!).

The other version number is the version of LiveCode in the release
note. This was erroneously set to 9.0.0-dp1, when in fact it should
be 9.1.0-dp1.

livecodeali commented:
@andrewferguson Great stuff :-)
@livecode-vulcan review ok c49e04e

livecode-vulcan commented:
💙 review by @livecodeali ok c49e04e3047792c81c622dd92bed450802bb41d8

livecode-vulcan commented:
😎 test success c49e04e3047792c81c622dd92bed450802bb41d8

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/61)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/53)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/53)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/53)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/53)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1575)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/102)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/102)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1605)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1599)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1595)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1632)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1572)
